### PR TITLE
chore(v2):  metastore API refactoring

### PR DESCRIPTION
**Pull request by kolesnikovae**
I decided to split https://github.com/grafana/pyroscope/pull/3652 into multiple PRs to avoid bad merge conflicts. In this PR, I refactor the FSM module and gRPC services. The core idea: strict raft command handler interface with storage/FSM injection (`*bbolt.Tx` for now). This required quite a few adjustments, but now we always perform all the operations in a single transaction and the state is explicitly provided to the command handler.

I tried to preserve as much as possible but decided to not re-write compactor tests as the implementation is changed significantly in https://github.com/grafana/pyroscope/pull/3652.

I'm going to test the PR more extensively (there almost no functional changes), but it is ready for review.

### CODE REVIEW

File: pkg/experiment/metastore/fsm/fsm.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+	if err = tx.Commit(); err != nil {
+		panic(fmt.Sprint("failed to commit transaction:", err))
+	}
+
+	return Response{Data: data, Err: err}
+}
+
+func (fsm *FSM) ReadOnlyTx() *bbolt.Tx {
+	tx, err := fsm.db.boltdb.Begin(true)
```
Comment (kolesnikovae) [inline pkg/experiment/metastore/fsm/fsm.go:151]:
```suggestion
	tx, err := fsm.db.boltdb.Begin(false)
```

Comment (kolesnikovae) [reply pkg/experiment/metastore/fsm/fsm.go:151]: read only txn

**Review by aleks-p — APPROVED** [review]
Looks great to me, thank you for taking the time to do the refactoring.

The only part I am not sure about is how clear some of the new concepts are. In particular, we have the `raft_node.Leader` and `raft_node.Follower` types. Their main purpose is ensuring read consistency but their names can be misleading because they are active on both leader and follower nodes. There is also a strong connection between the two, with the follower being the only consumer of what the leader provides (`ReadIndex`). I wonder if merging them under a `StateReader` (or `StateProvider`) is possible and could make their purpose more clear.

File: pkg/experiment/metastore/metastore.go [hunk 2]
```diff
@@ -109,41 +111,51 @@ func (cfg *RaftConfig) Validate() error {
...
 
-	walDir string
+	// Local state machine.
+	fsm *fsm.FSM
+	// An interface to make proposals.
+	proposer *RaftProposer
+	client   metastorev1.RaftNodeServiceClient
+	observer *raft_node.Observer
+	follower *raft_node.Follower
+	leader   *raft_node.Leader
```
[the remark this answers is not in the document]
Comment (kolesnikovae) [reply pkg/experiment/metastore/metastore.go:137]: I'm thinking about refactoring raft-related stuff to a separate module: now the metastore type has too many raft-somthing members. Will do that in the PR I'm currently working on

File: pkg/experiment/metastore/metastore.go [hunk 3]
```diff
@@ -154,62 +166,90 @@ func New(
...
+		m.cleanerHandler.CleanBlocks)
+
+	m.dlqRecovery = dlq.NewRecovery(logger, config.DLQRecovery, m.indexService, bucket)
+
+	// These are the services that only run on the raft leader.
+	// Keep in mind that the node may not be the leader at the moment the
+	// service is starting, so it should be able to handle conflicts.
+	m.observer.OnLeader(m.cleanerService)
+	m.observer.OnLeader(m.dlqRecovery)
+	m.observer.OnLeader(m.placement)
```
[the remark this answers is not in the document]
Comment (kolesnikovae) [reply pkg/experiment/metastore/metastore.go:229]: WDYT about `RunOnLeader`? We already have quite a few "registrations", and adding some more might not be helpful for the reader

File: pkg/experiment/metastore/raft_node/observer.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+		_, ok := o.Data.(raft.RaftState)
+		return ok
+	})
+	r.RegisterObserver(o.observer)
+	o.updateRaftState()
+	go o.run()
+	return o
+}
+
+func (o *Observer) RegisterHandler(h StateHandler) {
```
[the remark this answers is not in the document]
Comment (kolesnikovae) [reply pkg/experiment/metastore/raft_node/observer.go:76]: `RegisterHandler` is actually the primary method. Additionally, I plan to extend the interface to handle notifications for other types of events, such as peer additions or removals, leader loss, and so on

**Comment by kolesnikovae** [comment]
Thank you so much for the review, Aleks!

> The only part I am not sure about is how clear some of the new concepts are. In particular, we have the `raft_node.Leader` and `raft_node.Follower` types. Their main purpose is ensuring read consistency, but their names can be misleading because they are active on both leader and follower nodes.

Interestingly, I actually find these types intuitive and quite like them: the types encapsulate behavior rather than role.

For example, I really like that `follower.ConsistentRead` might be called on the leader (keep in mind that followers may be ahead of the leader in terms of applied changes, so the leader must also perform this check). This makes the Follower Reads pattern quite obvious.

Similarly, `leader.ReadIndex` can be called on any node: if the node is not elected by the quorum, the method fails; otherwise, it succeeds.

> There is also a strong connection between the two, with the follower being the only consumer of what the leader provides (`ReadIndex`). I wonder if merging them under a `StateReader` (or `StateProvider`) is possible and could make their purpose more clear.

You've clearly described the two components, how they interact, and the principle behind their connection: they indeed have a strong leader-follower connection and distinct responsibilities. I wouldn't want to obscure this from the user; in the current implementation, each component has a single responsibility.

But I hear you – I'll try to find a better better approach. I'm not entirely sure about `stateReader.ReadIndex` (vs. `leader.ReadIndex`) or `stateProvider.ConsistentRead` (vs. `follower.ConsistentRead`), but I'll give it a try.

**Pull request merged**